cranelift/codegen/src/opts/arithmetic.isle

```diff
@@ -343,3 +343,17 @@
 ;; (x + y) - y --> x
 (rule (simplify (isub ty (iadd ty x y) x)) y)
 (rule (simplify (isub ty (iadd ty x y) y)) x)
+
+;; (x * y) (==/!=) z --> x (==/!=) (y / z) when y is odd and divides z
```

Comment:
Could you add "and y and z are constant" to the description here? Otherwise this reads as a fairly odd optimization that is replacing a multiply with a divide, which would ordinarily reduce performance.

Also: why is it necessary that `y` is odd? Is that because we need it to be mutually prime with the ring's modulus (2^n) so the product doesn't collapse to zero?

Reply:
I've changed the description so that one can read that the division happens in compile time.
Also,  the solution X for X * C = D is not unique when C is even in 2^n module arithmetic.
Therefore, it is not safe to say X = D / C